cibyl/orchestrator.py

```diff
@@ -386,27 +395,27 @@ def query():
                 for source in system.sources:
                     source.ensure_teardown()
 
-        def publish():
-            self.publisher.publish(
-                environment=env, target=target, style=output_style,
-                query=query_type,
-                verbosity=self.parser.app_args.get('verbosity', 0),
-                **kwargs)
-
         command = self.parser.app_args.get('command')
         query_type = get_query_type(**self.parser.ci_args, command=command)
-        kwargs = {}
 
         target = PublisherTarget.TERMINAL
+        file = None
         if output_path:
             target = PublisherTarget.FILE
 
             # Overwrite file if it already exists
             file = File(output_path, resolve_home)
             file.delete()
             file.create()
-            kwargs['output_path'] = file
+
+        self.publisher = PublisherFactory.create_publisher(
+                target=target,
+                style=output_style,
+                query=query_type,
+                verbosity=self.parser.app_args.get('verbosity', 0),
+                output_file=file)
 
         for env in self.environments:
             query()
-            publish()
+            self.publisher.publish(environment=env)
+        self.publisher.finish_publishing()
```

Comment:
Overall, I think that the concept behind the publisher class is kind of outdated. In the end, it does not do much more than appending output from the printers. I think that all of this can be simplified with the IndentedTextBuilder tool and a couple of if statements. For example, these could be swapped for:
```
output = Indented...()

for ...:
    output.add(printer.print_env(env), 0)

output.build()
```

Reply:
I have tried your suggestion and you're right, it can be done (I have uploaded a branch with the change to my fork, [link](https://github.com/cescgina/cibyl/blob/d3fb2e6103df41c21d6aa9e6dedd34e8b7b4f5fe/cibyl/orchestrator.py#L413)). It's somewhat simpler, but I'm not sure if it would be simpler in the long run. If for example in the future we want to support yaml files, we will need to add another branch to the if, and then I'm not sure we're actually better off.

Reply:
Ok, let's go with this one then.